server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/model/MailboxFactory.scala

```diff
@@ -33,48 +33,38 @@ import scala.jdk.CollectionConverters._
 import scala.jdk.OptionConverters._
 
 object MailboxValidation {
-  private val errors: Seq[String] = Seq()
 
-  private def validateMailboxName(mailboxName: Either[String, MailboxName]): Option[MailboxName] = mailboxName match {
-    case Left(error) => {
-      errors.+(error)
-      None
-    }
-    case scala.Right(value) => Some(value)
+  private def validateMailboxName(mailboxName: Either[IllegalArgumentException, MailboxName]): MailboxName = mailboxName match {
+    case Left(error) => throw error
+    case scala.Right(value) => value
   }
 
-  private def validateNumber(number: Either[String, UnsignedInt]): Option[UnsignedInt] = number match {
-    case Left(error) => {
-      errors.+(error)
-      None
-    }
-    case scala.Right(value) => Some(value)
+  private def validateNumber(number: Either[NumberFormatException, UnsignedInt]): UnsignedInt = number match {
+    case Left(error) => throw error
+    case scala.Right(value) => value
   }
 
-  def validate(mailboxName: Either[String, MailboxName],
-               unreadEmails: Either[String, UnsignedInt],
-               unreadThreads: Either[String, UnsignedInt],
-               totalEmails: Either[String, UnsignedInt],
-               totalThreads: Either[String, UnsignedInt]): Either[Seq[String], MailboxValidation] = {
-    val validatedName: Option[MailboxName] = validateMailboxName(mailboxName)
+  def validate(mailboxName: Either[IllegalArgumentException, MailboxName],
+               unreadEmails: Either[NumberFormatException, UnsignedInt],
+               unreadThreads: Either[NumberFormatException, UnsignedInt],
+               totalEmails: Either[NumberFormatException, UnsignedInt],
+               totalThreads: Either[NumberFormatException, UnsignedInt]): MailboxValidation = {
+    val validatedName: MailboxName = validateMailboxName(mailboxName)
```

Comment:
all this method should be rewrote with for comprehension.

```
for {
  validatedName <- mailboxName
  validatedUnreadEmails <- unreadEmails.map(UnreadEmails(_))
 ...
} yield(MailboxValidation(
 mailboxName = validatedName,
 ...
))
```

which will return an Either[Exception, MailboxValidation] which is IHMO what this method should return.
if you want to accumulate the errors and not fail fast. You could use the Validation class in the cats library.
But i don't know if everybody here is ok to use it.

Reply:
On a side note, maybe you should define your own AST for validation errors.
creating an exception is expensive as it capture a whole stacktrace.